Wrappers/Python/ccpi/optimisation/Algorithms.py

```diff
@@ -0,0 +1,304 @@
+# -*- coding: utf-8 -*-
+#   This work is part of the Core Imaging Library developed by
+#   Visual Analytics and Imaging System Group of the Science Technology
+#   Facilities Council, STFC
+
+#   Copyright 2019 Edoardo Pasca
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#       http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+import numpy
+import time
+from ccpi.optimisation.funcs import ZeroFun
+
+class Algorithm(object):
+    '''Base class for iterative algorithms
+
+      provides the minimal infrastructure.
+      Algorithms are iterables so can be easily run in a for loop. They will
+      stop as soon as the stop cryterion is met.
+      The user is required to implement the set_up, __init__, update and
+      should_stop and update_objective methods
+   '''
+
+    def __init__(self):
+        self.iteration = 0
+        self.stop_cryterion = 'max_iter'
+        self.__max_iteration = 0
+        self.__loss = []
+        self.memopt = False
+        self.timing = []
+    def set_up(self, *args, **kwargs):
+        raise NotImplementedError()
+    def update(self):
+        raise NotImplementedError()
+    
+    def should_stop(self):
+        '''stopping cryterion'''
+        raise NotImplementedError()
+    
+    def __iter__(self):
+        return self
+    def next(self):
+        '''python2 backwards compatibility'''
+        return self.__next__()
+    def __next__(self):
+        if self.should_stop():
+            raise StopIteration()
+        else:
+            time0 = time.time()
+            self.update()
+            self.timing.append( time.time() - time0 )
+            self.update_objective()
+            self.iteration += 1
+    def get_output(self):
+        '''Returns the solution found'''
+        return self.x
+    def get_current_loss(self):
```

Comment:
Suggest to rename `get_current_loss` as `get_objective`

Reply:
I added an alias, so `get_current_objective` and `get_current_loss` are both available